patches/tModLoader/Terraria/Social/Steam/WorkshopSocialModule.TML.cs

```diff
@@ -189,6 +202,54 @@ private bool _PublishMod(TmodFile modFile, NameValueCollection buildData, Worksh
 		return false;
 	}
 
+	// PR 4345 - We combine the hash data that is currently on workshop with the hash data from the updated publishing folder to ensure that when mods are updated it is backwards compatible
+	// It is backwards compatible while Steam spends up to an hour rolling out workshop item updates
+	private static void CalculateModHashes(string workshopPath, ModDownloadItem modDownloadItemAsFound, ref DeveloperMetadata devMetadata)
+	{
+		// Get the hashes from the existing modDownloadItem as found on the workshop
+		var prevHashes = modDownloadItemAsFound.GetModVersionHashes();
+
+		// Get the new hashes
+		var currentHashes = new List<ModVersionHash>();
+		foreach (var tModPath in Directory.EnumerateFiles(workshopPath, "*.tmod*", SearchOption.AllDirectories)) {
+			var tModFile = new TmodFile(tModPath);
+			tModFile.Open(); // Needed for Hash data to be populated
+			currentHashes.Add(new ModVersionHash(tModFile));
+		}
+
+		List<ModVersionHash> totalHash = currentHashes.Concat(prevHashes.Except(currentHashes).ToList()).ToList();
+
+		devMetadata.modVersionHashes = totalHash.Select(h => h.ToString()).ToList();
+	}
+
+	// This methods trims contents of developer metadata based on the preferred order of discarding information.
+	// It is primarily written with the intent of 'in case' we need to store other information in this Workshop text field
+	private static void TrimDevMetadata(ref DeveloperMetadata devMetadata)
+	{
+		const int MaxMetadataLength = Steamworks.Constants.k_cchDeveloperMetadataMax;
+
+		var overflowLength = JsonConvert.SerializeObject(devMetadata).Length - MaxMetadataLength;
+		if (overflowLength <= 0)
+			return;
+
+		int charsSaved = 0;
+
+		// Check if we can reduce the number of ModHashes
+		var minNumberOfHashes = 2 * SocialBrowserModule.keepRequirements.Select(a => a.keepCount).Sum();
```

Comment:
Surely this whole function could be simpler if we just ran a while loop and dropped the oldest hash and reserialized. There's only ever going to be a small handful of hashes to drop, and serialization takes a negligible amount of time. 

We can check that the number of hashes we kept at the end is at least the number in the current upload for your exception below